localstack-core/localstack/services/cloudwatch/provider_v2.py

```diff
@@ -454,8 +454,18 @@ def put_composite_alarm(self, context: RequestContext, request: PutCompositeAlar
             composite_alarm = LocalStackCompositeAlarm(
                 context.account_id, context.region, {**request}
             )
+
+            alarm_rule = composite_alarm.alarm["AlarmRule"]
+
+            alarms_from_rule = [alarm.strip() for alarm in alarm_rule.split("OR")]
+            for alarm in alarms_from_rule:
+                if not alarm.startswith("ALARM"):
+                    raise ValidationError(
+                        f"Error in alarm rule condition '{alarm}': Only ALARM expression is supported"
+                    )
```

Comment:
I think we should perhaps `LOG.warning` and ignore the rule rather than raising an error on unsupported actions. This becomes a known lack of parity rather than a user error.

... _except_ if this is actually a parity improvement with AWS, in which case we should have a snapshot to cover it. I think this would be better as an additional test to capture any validation errors thrown by CloudWatch.

Reply:
I did take inspiration from AWS response for `put-composite-alarm` when the alarm rule grammar has unsupported tokens.
For example, if I run the following command against AWS

```aws cloudwatch put-composite-alarm --alarm-name "composite-alarm" --alarm-rule "ALARRRRM(metric-alarm)"```

then the response is:

```An error occurred (ValidationError) when calling the PutCompositeAlarm operation: Error in AlarmRule [Unsupported token 'ALARRRRM' at char 0]```

I'm not sure if to qualify the validation I added as a parity improvement though. My motivation to add this validation was to make it very explicit to a Localstack user that as of now no expressions except `ALARM` one are supported. This way when they try to add a composite alarm that has expressions that are not supported yet by Localstack they will get a message and error response quite similar to what AWS provides. If there is no such validation a user could add a composite alarm with what is supported by AWS, let's say, `OK(metric-alarm)` in the rule. This rule would be saved and later be evaluated not as user expects, because in the current implementation every metric alarm is checked for ALARM state only.

So what I added is a _similar_ behaviour, not full parity.

Because of that, I didn't see how I can do a shapshot, as AWS would happily accept the `ALARM` token.

I think when a full alarm rule grammar is implemented many things will become simpler and cleaner, validation included. I just didn't venture into implementing full grammar parser in the first iteration, which lead to having to make several workarounds.

After my explanation, do you think it is worth to keep this validation? I'm absolutely good to change to `LOG.warning` as you suggest if that's the go-to approach in Localstack for lack of parity.

Reply:
I agree a line needs to be drawn for this first implementation. My take is: I would prefer to not fail on _valid_ input if it is not supported by LocalStack (yet). I think it is better to not break existing infrastructure (e.g. if a user tries to deploy their app to LocalStack, we should not block them just because we don't support a feature).

The only other option we have is to raise `NotImplementedError` which correctly conveys the message, however we treat that exception specially to return a response similar to "The API action '...' is not supported" which is not quite what we want. The _action_ is supported, just that this specific input is not.

Can we maybe switch to the `LOG` statement in this instance, and as you say: a proper rule parser will likely be a future step and we can properly validate the input at that stage.

Reply:
> I think it is better to not break existing infrastructure (e.g. if a user tries to deploy their app to LocalStack, we should not block them just because we don't support a feature).

Great point, now what you suggest makes a lot of sense to me! Also it made me think whether a partial rule implementation does more good than harm. 😄 I mean situations like when someone has a rule with AND in their infra and Localstack will treat it as if it was OR. Might be a rare case, yet it is probably better to not delay full alarm expression grammar support by a long time after (if) this gets merged.

Anyway, will now change raising error to logging as you suggest, thanks for explaining the details about parity concept!